nixos/modules/services/mail/listmonk.nix

```
@@ -0,0 +1,212 @@
+                  };
+                  tls_type = mkOption {
+                    type = types.enum [ "none" "STARTTLS" "TLS" ];
+                    description = lib.mdDoc "Type of TLS authentication with the SMTP server";
+                  };
+                };
+              });
+
+              description = lib.mdDoc "List of outgoing SMTP servers";
+            };
+
+            # TODO: refine this type based on the smtp one.
+            "bounce.mailboxes" = mkOption {
+              type = listOf (submodule {
+                freeformType = with types; oneOf [ str int bool ];
+              });
+              default = [];
+              description = lib.mdDoc "List of bounce mailboxes";
+            };
+
+            messengers = mkOption {
+              type = listOf str;
+              default = [];
+              description = lib.mdDoc "List of messengers, see: <https://github.com/knadh/listmonk/blob/master/models/settings.go#L64-L74> for options.";
+            };
+          };
+        });
+
+        description = lib.mdDoc "Dynamic settings in the PostgreSQL database, set by a SQL script, see <https://github.com/knadh/listmonk/blob/master/schema.sql#L177-L230> for details.";
+      };
+      secretFile = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = lib.mdDoc "Secret file loaded as an environment file, read <https://listmonk.app/docs/configuration/#environment-variables> for details.";
+      };
+    };
+  };
+
+
+  ###### implementation
+  config = mkIf cfg.enable (mkMerge [
+    {
+      services.listmonk.settings."app".address = mkDefault "localhost:9000";
+
+      # Default parameters from https://github.com/knadh/listmonk/blob/master/config.toml.sample
+      services.listmonk.settings = {
+        "db".max_open = mkDefault 25;
+        "db".max_idle = mkDefault 25;
+        "db".max_lifetime = mkDefault "300s";
+      };
+
+      users.users."listmonk" = {
+        isSystemUser = true;
+        group = "listmonk";
+      };
+      users.groups."listmonk" = {};
+
+      systemd.services.listmonk = {
+        description = "Listmonk - newsletter and mailing list manager";
+        after = [ "network.target" ] ++ optional cfg.database.createLocally "postgresql.service";
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          Type = "exec";
+          EnvironmentFile = if (cfg.secretFile != null) then [ cfg.secretFile ] else [];
+          ExecStartPre = [
+            ''${pkgs.coreutils}/bin/echo "$STATE_DIRECTORY/uploads"; ${pkgs.coreutils}/bin/mkdir -p "$STATE_DIRECTORY/uploads"''
+            "${cfg.package}/bin/listmonk --config ${cfgFile} --idempotent --install --upgrade --yes"
+            "${updateDatabaseConfigScript}/bin/update-database-config.sh"
+          ];
+          ExecStart = "${cfg.package}/bin/listmonk --config ${cfgFile}";
+
+          Restart = "on-failure";
+
+          StateDirectory = "listmonk";
```

```suggestion
          StateDirectory = [ "listmonk" "listmonk/uploads" ];
```

Avoid the above `mkdir` command i favour of `StateDirectory`.

---

Addressed.